Lib/shutil.py

```diff
@@ -1550,21 +1550,21 @@ def which(cmd, mode=os.F_OK | os.X_OK, path=None):
     if sys.platform == "win32":
         # PATHEXT is necessary to check on Windows.
         pathext_source = os.getenv("PATHEXT") or _WIN_DEFAULT_PATHEXT
-        pathext = [ext for ext in pathext_source.split(os.pathsep) if ext]
+        pathext = pathext_source.split(os.pathsep)
+        pathext = [ext.rstrip('.') and ext for ext in pathext if ext]
```

Comment:
I really dislike using `and` for value selection. I trust that you've got it right here, but I can't read it easily. Can we maybe comment?

```
# Replace ext of '.' with empty string, but keep trailing dots on '.ext.'
```

Reply:
Actually, I think that it is better to remove dots in all cases. The behavior of cmd.exe if `PATHEXT` contains `.BAT.` and there is `a.bat` in the path:

* `a` -- hangs
* `a.bat` -- found
* `a.` -- not found
* `a.bat.` -- not found

`which()` will now return the same result, except for the first case. I think this is the best option for now.

Reply:
Sounds good to me